src/dex/woo-fi/woo-fi.ts

```diff
@@ -0,0 +1,556 @@
+    }
+    return this._exchangeAddress;
+  }
+
+  get oracleAddress() {
+    if (this._oracleAddress === undefined) {
+      this._oracleAddress = this.config.woOracleAddress.toLowerCase();
+    }
+    return this._oracleAddress;
+  }
+
+  get quoteTokenAddress() {
+    if (this._quoteTokenAddress === undefined) {
+      this._quoteTokenAddress = this.config.quoteToken.address.toLowerCase();
+    }
+    return this._quoteTokenAddress;
+  }
+
+  protected _fillTokenInfoState(
+    state: PoolState,
+    address: string,
+    values: Result,
+  ) {
+    state.tokenInfos[address.toLowerCase()] = {
+      reserve: BigInt(values.reserve._hex),
+      R: BigInt(values.R._hex),
+      threshold: BigInt(values.threshold._hex),
+    };
+  }
+
+  async fetchStateForBlockNumber(blockNumber?: number): Promise<PoolState> {
+    const calldata = _.flattenDeep(
+      this.baseTokens.map(t => [
+        {
+          target: this.config.wooFeeManagerAddress,
+          callData: this.wooIfaces.fee.encodeFunctionData('feeRate', [
+            t.address,
+          ]),
+        },
+        {
+          target: this.config.woOracleAddress,
+          callData: this.wooIfaces.oracle.encodeFunctionData('infos', [
+            t.address,
+          ]),
+        },
+        {
+          target: this.config.wooPPAddress,
+          callData: this.wooIfaces.PP.encodeFunctionData('tokenInfo', [
+            t.address,
+          ]),
+        },
+      ]),
+    );
+
+    calldata.push({
+      target: this.config.wooPPAddress,
+      callData: this.wooIfaces.PP.encodeFunctionData('tokenInfo', [
+        this.quoteTokenAddress,
+      ]),
+    });
+
+    const data = await this.dexHelper.multiContract.methods
+      .aggregate(calldata)
+      .call({}, blockNumber || 'latest');
+
+    // Last request is standalone
+    const maxNumber = calldata.length - 1;
+
+    const [baseFeeRates, baseInfos, baseTokenInfos, quoteTokenInfo] = [
+      // Skip two as they are infos abd tokenInfo
+      _.range(0, maxNumber, 3).map(index =>
+        this.wooIfaces.fee.decodeFunctionResult(
+          'feeRate',
+          data.returnData[index],
+        ),
+      ),
+      _.range(1, maxNumber, 3).map(index =>
+        this.wooIfaces.oracle.decodeFunctionResult(
+          'infos',
+          data.returnData[index],
+        ),
+      ),
+      _.range(2, maxNumber, 3).map(index =>
+        this.wooIfaces.PP.decodeFunctionResult(
+          'tokenInfo',
+          data.returnData[index],
+        ),
+      ),
+      this.wooIfaces.PP.decodeFunctionResult(
+        'tokenInfo',
+        data.returnData[maxNumber],
+      ),
+    ];
+
+    const state: PoolState = {
+      feeRates: {},
+      tokenInfos: {},
+      tokenStates: {},
+    };
+    this._fillTokenInfoState(state, this.quoteTokenAddress, quoteTokenInfo);
+
+    baseFeeRates.map((value, index) => {
+      const tokenAddress = this.baseTokens[index].address.toLowerCase();
```

Comment:
if you follow my comment above you don't need to lower case anymore

Reply:
Fixed